synapse/handlers/federation_event.py

```diff
@@ -766,10 +766,21 @@ async def _process_pulled_event(
         """
         logger.info("Processing pulled event %s", event)
 
-        # these should not be outliers.
-        assert (
-            not event.internal_metadata.is_outlier()
-        ), "pulled event unexpectedly flagged as outlier"
+        # This function should not be used to persist outliers. If you happen to
+        # run into a situation where the event you're trying to process/backfill
+        # is marked as an `outlier`, then you should update that spot to return
+        # an `EventBase` copy that doesn't have `outlier` flag set.
+        #
+        # `EventBase` is used to represent both an event we have not yet
+        # persisted, and one that we have persisted and now keep in the cache.
+        # In an ideal world this method would only be called with the first type
+        # of event, but it turns out that's not actually the case and for
+        # example, you could get an event from cache that is marked as an
+        # `outlier` (fix up that spot though).
+        assert not event.internal_metadata.is_outlier(), (
+            "This is a safe-guard to make sure we're not trying to persist an outlier using this function (use something else). "
+            "If you're trying to process/backfill an event, this is the right method but you need pass in an event copy that doesn't have `event.internal_metada.outlier = true`."
```

Comment:
Well, if your goal is to avoid having to read the code, then you better say *which* function, rather than just "this function".

I think one objection I have here is that it's extremely wordy. For example: `This is a safe-guard to make sure we're not trying to persist an outlier using [_process_pulled_event] (use something else)` contains *exactly* the same amount of information as `outlier event passed to _process_pulled_event`, imho, but says it in a much less clear way.

I'm also not convinced the words are correct.  "If you're trying to process/backfill an event, this is the right method": well, maybe, but it depends what sort of processing you're trying to do. If you're trying to persist an outlier, it's certainly not the right method.

(also: `metada`.) 

How about this:

```
Outlier event passed to _process_pulled_event. To persist an event as a non-outlier, make sure to pass in a copy without `event.internal_metadata.outlier = true`.
```

Reply:
My goal is for the person to know what to do to fix it. "this function" and the stack to get there is already visible in the `AssertionError` output (see "Log example")

---

> How about this:
> 
> ```
> Outlier event passed to _process_pulled_event. To persist an event as a non-outlier, make sure to pass in a copy without `event.internal_metadata.outlier = true`.
> ```

This is a great suggestion which still addresses my goal 👍 Updated ⏩

Thanks for working through this with me!